numba/cuda/random.py

```diff
@@ -0,0 +1,281 @@
+from __future__ import print_function, absolute_import
+import math
+
+from numba import cuda, float32, float64, uint32, int64, uint64, from_dtype,\
+                  jit
+
+import numpy as np
+
+# This implementation is based upon the xoroshiro128+ and splitmix64 algorithms
+# described at:
+#
+#     http://xoroshiro.di.unimi.it/
+#
+# and originally implemented by David Blackman and Sebastiano Vigna.
+#
+# The implementations below are based on the C source code:
+#
+#  * http://xoroshiro.di.unimi.it/xoroshiro128plus.c
+#  * http://xoroshiro.di.unimi.it/splitmix64.c
+#
+# Splitmix64 is used to generate the initial state of the xoroshiro128+
+# generator to ensure that small seeds don't result in predictable output.
+
+# **WARNING**: There is a lot of verbose casting in this file to ensure that
+# NumPy casting conventions (which cast uint64 [op] int32 to float64) don't
+# turn integers into floats when using these functions in the CUDA simulator.
+#
+# There are also no function type signatures to ensure that compilation is
+# deferred so that import is quick, and Sphinx autodoc works.  We are also
+# using the CPU @jit decorator everywhere to create functions that work as
+# both CPU and CUDA device functions.
+
+xoroshiro128p_dtype = np.dtype([('s0', np.uint64), ('s1', np.uint64)])
```

Comment:
Use align struct here: `np.dtype([('s0', np.uint64), ('s1', np.uint64)], align=True)`.
Otherwise, unalighed load is 4x slower:

```
Time(%)      Time     Calls       Avg       Min       Max  Name
 99.95%  41.157ms         1  41.157ms  41.157ms  41.157ms  cudapy::numba::cuda::random::init_xoroshiro128p_states_kernel$241(Array<Record<int=177>, int=1, A, mutable, unaligned>, __int64, __int64)
```
```
Time(%)      Time     Calls       Avg       Min       Max  Name
 99.90%  13.570ms         1  13.570ms  13.570ms  13.570ms  cudapy::numba::cuda::random::init_xoroshiro128p_states_kernel$241(Array<Record<int=177>, int=1, A, mutable, aligned>, __int64, __int64)
```